node/cmd/spy/spy.go

```diff
@@ -69,84 +69,244 @@ var SpyCmd = &cobra.Command{
+				sub.ch <- message{vaaBytes: vaaBytes}
+			}
+		}
+
 	}
-	if len(address) != 32 {
-		return vaa.Address{}, status.Error(codes.InvalidArgument, "address must be 32 bytes")
+
+	return nil
+}
+
+// TransactionIdMatches checks if both TxIds have the same value.
+func TransactionIdMatches(g *gossipv1.SignedBatchVAAWithQuorum, t *spyv1.BatchFilter) bool {
+	return bytes.Equal(g.TxId, t.TxId)
+}
+
+// BatchMatchFilter asserts that the obervation matches the values of the filter.
+func BatchMatchesFilter(g *gossipv1.SignedBatchVAAWithQuorum, f *spyv1.BatchFilter) bool {
+	// check the chain ID
+	if g.ChainId != uint32(f.ChainId) {
+		return false
+	}
+
+	// check the transaction ID
+	txMatch := TransactionIdMatches(g, f)
+	if !txMatch {
+		return false
 	}
 
-	addr := vaa.Address{}
-	copy(addr[:], address)
+	// check the Nonce
+	if f.Nonce >= 1 {
+		// filter has a nonce, so make sure it matches
+		if g.Nonce != f.Nonce {
+			// filter's nonce does not match the nonce of the Batch.
+			return false
+		}
+	}
 
-	return addr, nil
+	return true
 }
 
-func (s *spyServer) Publish(vaaBytes []byte) error {
-	s.subsMu.Lock()
-	defer s.subsMu.Unlock()
+// HandleGossipVAA compares a gossip message to client subscriptions & filters,
+// and forwards the VAA to those requesting it.
+func (s *spyServer) HandleGossipVAA(g *gossipv1.SignedVAAWithQuorum) error {
+	s.subsAllVaaMu.Lock()
+	defer s.subsAllVaaMu.Unlock()
 
-	var v *vaa.VAA
+	v, err := vaa.Unmarshal(g.Vaa)
+	if err != nil {
+		s.logger.Error("failed unmarshaing VAA bytes from gossipv1.SignedVAAWithQuorum.",
+			zap.Error(err))
+		return err
+	}
+
+	// resType defines which oneof proto will be retuned - res type "SignedVaa" is *gossipv1.SignedVAAWithQuorum
+	resType := &spyv1.SubscribeSignedVAAByTypeResponse_SignedVaa{
+		SignedVaa: g,
+	}
+
+	// envelope is the highest level proto struct, the wrapper proto that contains one of the VAA types.
+	envelope := &spyv1.SubscribeSignedVAAByTypeResponse{
+		VaaType: resType,
+	}
 
-	for _, sub := range s.subs {
+	// loop through the subscriptions and send responses to everyone that wants this VAA
+	for _, sub := range s.subsAllVaa {
 		if len(sub.filters) == 0 {
-			sub.ch <- message{vaaBytes: vaaBytes}
-		} else {
-			if v == nil {
-				var err error
-				v, err = vaa.Unmarshal(vaaBytes)
-				if err != nil {
-					return err
+			// this subscription has no filters, send them the VAA.
+			sub.ch <- envelope
+			continue
+		}
+
+		// this subscription has filters.
+		for _, filterEntry := range sub.filters {
+			filter := filterEntry.GetFilter()
+			switch t := filter.(type) {
+			case *spyv1.FilterEntry_EmitterFilter:
+				filterAddr := t.EmitterFilter.EmitterAddress
+				filterChain := vaa.ChainID(t.EmitterFilter.ChainId)
+
+				if v.EmitterChain == filterChain && v.EmitterAddress.String() == filterAddr {
+					// it is a match, send the response
+					sub.ch <- envelope
 				}
+			default:
+				panic("unsupported filter type in subscriptions")
```

Comment:
might also want to consider including the type here (via %T) to help with debugging:

```suggestion
				panic("unsupported filter type in subscriptions: %T", filter)
```

Reply:
huh, TIL. added it, thanks.